cmd/system-probe/subcommands/run/command.go

```diff
@@ -141,25 +141,73 @@ func run(log log.Component, config config.Component, telemetry telemetry.Compone
+
 	// run startSystemProbe in an app, so that the log and config components get initialized
-	return fxutil.OneShot(
-		func(log log.Component, config config.Component, telemetry telemetry.Component, sysprobeconfig sysprobeconfig.Component, rcclient rcclient.Component) error {
-			return startSystemProbe(&cliParams{GlobalParams: &command.GlobalParams{}}, log, telemetry, sysprobeconfig, rcclient)
-		},
-		// no config file path specification in this situation
-		fx.Supply(config.NewAgentParamsWithoutSecrets("", config.WithConfigMissingOK(true))),
-		fx.Supply(sysprobeconfig.NewParams(sysprobeconfig.WithSysProbeConfFilePath(""))),
-		fx.Supply(log.LogForDaemon("SYS-PROBE", "log_file", common.DefaultLogFile)),
-		rcclient.Module,
-		config.Module,
-		telemetry.Module,
-		sysprobeconfig.Module,
-		// use system-probe config instead of agent config for logging
-		fx.Provide(func(lc fx.Lifecycle, params log.Params, sysprobeconfig sysprobeconfig.Component) (log.Component, error) {
-			return log.NewLogger(lc, params, sysprobeconfig)
-		}),
-	)
+	go func() {
+		err := fxutil.OneShot(
+			func(log log.Component, config config.Component, telemetry telemetry.Component, sysprobeconfig sysprobeconfig.Component, rcclient rcclient.Component) error {
+				defer StopSystemProbeWithDefaults()
+				err := startSystemProbe(&cliParams{GlobalParams: &command.GlobalParams{}}, log, telemetry, sysprobeconfig, rcclient)
+				// notify outer that startAgent finished
+				errChan <- err
+				if err != nil {
+					return err
+				}
+				// wait for context
+				ctx := <-ctxChan
+
+				// Wait for stop signal
+				select {
+				case <-signals.Stopper:
+					log.Info("Received stop command, shutting down...")
+				case <-signals.ErrorStopper:
+					_ = log.Critical("The Agent has encountered an error, shutting down...")
+				case <-ctx.Done():
+					log.Info("Received stop from service manager, shutting down...")
+				}
+
+				return nil
+			},
+			// no config file path specification in this situation
+			fx.Supply(config.NewAgentParamsWithoutSecrets("", config.WithConfigMissingOK(true))),
+			fx.Supply(sysprobeconfig.NewParams(sysprobeconfig.WithSysProbeConfFilePath(""))),
+			fx.Supply(log.LogForDaemon("SYS-PROBE", "log_file", common.DefaultLogFile)),
+			rcclient.Module,
+			config.Module,
+			telemetry.Module,
+			sysprobeconfig.Module,
+			// use system-probe config instead of agent config for logging
+			fx.Provide(func(lc fx.Lifecycle, params log.Params, sysprobeconfig sysprobeconfig.Component) (log.Component, error) {
+				return log.NewLogger(lc, params, sysprobeconfig)
+			}),
+		)
+		// notify caller that fx.OneShot is done
+		errChan <- err
+	}()
+
+	// Wait for startSystemProbe to complete, or for an error
+	err := <-errChan
+	if err != nil {
+		// startSystemProbe or fx.OneShot failed, caller does not need errChan
+		return nil, err
+	}
+
+	// startSystemProbe succeeded. provide errChan to caller so they can wait for fxutil.OneShot to stop
+	return errChan, nil
```

Comment:
Won't the error sent to `errChan` already be consumed by the above code? How would the caller use the returned channel?